docs/src/docs/asciidoc/developer/clustered-events.adoc

```diff
@@ -0,0 +1,90 @@
+= Ehcache events from clustered caches
+
+This document describes the core design of what is required to implement ehcache events from caches backed by a
+clustered store.
+
+== High-level requirements
+
+* Ehcache supports five types of cache events: on *eviction*, on *expiry*, on *removal*, on *update* and on *creation*.
+* When an event is fired, every connected client with a registered listener has to receive it.
+* Events must be delivered once and only once as long as the client(s), server(s) and network-in-between are all healthy.
+* What happens when there is a client disconnect, a passive take over, a split brain or any other hazard is yet to be
+determined.
+* No performance impact when the feature isn't used.
+
+=== Recommandations
+
+It must be made clear (documentation?) that the eventing mechanism is going to have a performance impact, sometimes
+a very noticeable one.
+
+Some features are undesirable because they are unlikely to be practical:
+
+* Synchronous events would require waiting for a round-trip to all clients before achieving a cache operation. This
+would pretty much make such cache unusably slow. Such config should throw *UnsupportedOperationException* when attempted.
+* Guaranteeing event delivery in all cases would require some form of stable store and a fairly complex and costly
+2-phase logic. This would also have an unsustainable performance impact. Instead, clients should have a way to figure out
+when such hazard happen to compensate for the possible lack of event delivery.
+
+== Technical facts
+
+Because of the current implementation:
+
+* Most events have to find their source from a client, more specifically from the clustered `Store` implementation:
+*expiry*, *removal*, *update* and *creation* are all client-triggered.
+* One event has to originate from the server: *eviction* is server-triggered.
+
+This means a cluster-wide listener mechanism has to be created with the following features:
+
+* Clients can register and unregister themselves.
+* Events can be fired from any client or any server.
+* Events have to be transported either from a server to all listening clients, or from one client to all clients
+(sometimes including the source client) via a server. AA topologies complicate client-sourced events transport as events
+have to be delivered only once.
+* Events can be fired from non-listening clients.
+
+== The transport mechanism
+
+An event delivery mechanism must be built to transport the events from the clients and servers to all clients registered
+as listeners. It requires the following:
+
+* API to (un)register a client as a listener
+* API for a client to fire an event
+* API for a server to fire an event
```

Comment:
hum... why does the server need to send events ? I thought only clients would be firing / receving events ?